### ISSUE TITLE
[Question] Is the locator.count() function supposed to wait for elements matching that locator before returning the count value?

### ISSUE THREAD

**Issue by mathewr94** [microsoft/playwright#14278]
Hi,

I've been trying to use the [locator.count](https://playwright.dev/docs/api/class-locator#locator-count) function in my typescript Playwright test and I've been fumbling around trying to figure out why I was getting a 0 value returned from the count function. As I was debugging, I searched with the exact locator that my test was using on the chromium dev tools and it was finding 5 elements.

The original code I was using was:
```
const numberOfDresses: number = await this.page.locator(DressesPage.dressesPrices).count();
```
which was returning 0.

And I had to change it to this in order for it to work:
```
await this.page.locator(DressesPage.dressesPrices).nth(1).waitFor();
const numberOfDresses: number = await this.page.locator(DressesPage.dressesPrices).count();
```

So it got me wondering... Is the locator.count() function supposed to wait for elements matching that locator before returning the count value? Or have I stumbled across a bug?

Thanks,
Mathew

**Comment by mxschmitt** [comment]
locator.count() is not wait, it does the amount of elements which are found of the time you issue the command.

To wait for elements, you should use the web-first assertions: `await expect(page.locator('.fooobar').toHaveCount(42)` which waits automatically until the condition is met: https://playwright.dev/docs/test-assertions#locator-assertions-to-have-count

**Comment by mathewr94** [comment]
No, I actually want to get the count of how many elements appear on the screen which match a given locator, without any prior knowledge of how many might appear. In my exact situation, x number of dresses will be displayed for sale on a webpage. At any given point, the merchant could add or remove dresses, and my goal is to get a count of all dresses displayed on the webpage and loop through them to find and select the most expensive one, then add it to the cart. For this scenario, a toHaveCount(x) or hypothetically a toBeGreaterCount(x) function would not be suitable. 

Essentially, I think it might be useful to have a count() function which waits for the elements to be displayed first, e.g. waitThenCount()

**Comment by yury-s** [comment]
How do you determine that no more elements matching the condition will appear on the page? You should wait for that condition and call `.count()` after that.

**Comment by mathewr94** [comment]
In the end, I just used what I was using:
```
await this.page.locator(DressesPage.dressesPrices).nth(1).waitFor();
const numberOfDresses: number = await this.page.locator(DressesPage.dressesPrices).count();
```

This question was really raised because I didn't understand how .count() works and I thought I might've encountered a bug in the function. For anyone who might find this from a Google search, just know that the .count() function will get a count without waiting for any elements that match your provided locator, so you have to wait first.

**Comment by noumanshahid12** [comment]
> 

For Example ?

**Comment by costag1982** [comment]
> Are you interested in toBeGreaterCount? Like if you don't know what to wait for, then either toBeGreaterCount could be a valid feature request or you use your approach with just waitFor which waits for at least one element.

I would also love this feature as stumbled across the exact same issue today

**Comment by alvaromartmart** [comment]
Same here, I think there are many valid uses cases for this. I would name it `toHaveCountGreaterThan()` (and corresponding `toHaveCountLessThan`)

**Comment by mxschmitt** [comment]
> Same here, I think there are many valid uses cases for this. I would name it `toHaveCountGreaterThan()` (and corresponding `toHaveCountLessThan`)

You can do the following instead: 

```js
    await expect.poll(async () => page.locator('div').count()).toBeGreaterThan(5);
```

**Comment by alvaromartmart** [comment]
> > Same here, I think there are many valid uses cases for this. I would name it `toHaveCountGreaterThan()` (and corresponding `toHaveCountLessThan`)
> 
> You can do the following instead:
> 
> ```js
>     await expect.poll(async () => page.locator('div').count()).toBeGreaterThan(5);
> ```

Didn't know about `expect.poll`, thanks!!
I'd rather have an out of the box matcher but this workaround is at least better than having to implement the polling mechanism.

**Comment by mxschmitt** [comment]
Feel free to file a feature request for it!

**Comment by amitchaudhary140** [comment]
The above solution worked for me.

**Comment by akwasin** [comment]
> ```
>     await expect.poll(async () => page.locator('div').count()).toBeGreaterThan(5);
> ```
or
> ```
> await expect(page.locator('.fooobar').toHaveCount(42)
> ```

I stumbled upon this aswell today but I don't think that these should be the way forward and I cant explain it but it seems to work sometimes while it doesnt work at all at other moments

Works:
```
const itemCount: number = await page.locator('[data-testid="menu-link"] li').count(); // returns correct amount
const randomInt: number = Math.floor(Math.random() * ((itemCount) - 1 + 1) + 1);
await page.click(`[data-testid="menu-link:nth-child${randomInt}"]`)
```

Doesnt work:
const itemsList: number = await page.locator('[data-testid="menu-link"]').count(); // returns 0
expect(itemsList).toBe(4);

**Comment by ChuhengH** [comment]
> In the end, I just used what I was using:
> 
> ```
> await this.page.locator(DressesPage.dressesPrices).nth(1).waitFor();
> const numberOfDresses: number = await this.page.locator(DressesPage.dressesPrices).count();
> ```
> 
> This question was really raised because I didn't understand how .count() works and I thought I might've encountered a bug in the function. For anyone who might find this from a Google search, just know that the .count() function will get a count without waiting for any elements that match your provided locator, so you have to wait first.

What do you do if the this.page.locator(DressesPage.dressesPrices) doesn't exist?
Calling  await this.page.locator(DressesPage.dressesPrices).nth(1).waitFor(); will be timeout.
The.count() should wait and returns 0 if  this.page.locator(DressesPage.dressesPrices) doesn't exist.
I think it's a playwright's bug.

**Comment by Zerotask** [comment]
Maybe a `wait_for_count()` could be the solution here.

**Comment by 0xIslamTaha** [comment]
For anyone landing here, I want to clarify the main issue, as I’m sure none of the presented solutions fully address it.

The core problem—similar to my own—is that we need to check for an element's presence. If the element exists, we execute logicA; if it doesn’t, we proceed with logicB. This means that in some runs, the element will be present, and in others, it won’t—both scenarios must be handled properly.

- We don’t need to assert its presence.
- We don’t want the test to fail if the element doesn’t exist.
- The element count is either 0 or 1, and based on that, we need to take different actions.

**The Issue**:
According to the documentation, the methods all(), count(), and isVisible() do not wait for the element to appear—they simply return whatever is present on the page at that moment.

In my case, the element does exist, but these methods didn’t wait for it, leading to incorrect results:

<img width="707" alt="Image" src="https://github.com/user-attachments/assets/288dba79-0a40-43ff-b548-ed56b06a1a24" />


The solution is, We need to explicitly wait for the element and wrap the check in a **try-catch block**. Here’s how I handled it:

```js
try {
        await duplicateFilterLocator.waitFor({ timeout: 5000});
        console.log(await duplicateFilterLocator.count()); // 1
        console.log(await duplicateFilterLocator.isVisible()) // 1
        console.log(await duplicateFilterLocator.all()) // 1 
      } catch {
        await filterActions.saveFilterWithName(page, duplicateFilterName); // will be run if count() is zero
      }
    });
```

This ensures that the element has enough time to appear before we check its state, preventing false negatives.
